Objects/setobject.c

```diff
@@ -209,6 +203,57 @@ set_add_entry(PySetObject *so, PyObject *key, Py_hash_t hash)
     return -1;
 }
 
+static int
+set_add_entry(PySetObject *so, PyObject *key, Py_hash_t hash)
+{
+    _Py_CRITICAL_SECTION_ASSERT_OBJECT_LOCKED(so);
+
+    /* Pre-increment is necessary to prevent arbitrary code in the rich
+       comparison from deallocating the key just before the insertion. */
+    Py_INCREF(key);
+
+    return set_add_entry_takeref(so, key, hash);
```

Comment:
I don't think the comment makes much sense here anymore. The rich comparison doesn't matter: we need to incref here because `set_add_entry_takeref` steals a reference and `set_add_entry` does not.

```suggestion
    return set_add_entry_takeref(so, Py_NewRef(key), hash);
```